### PR TITLE
Updating the SDK with 0320 functionality

### PR DESCRIPTION
**Pull request by cmavromichalis**
*Added support for getting and setting the trust state for the TDI service from /api/devicetrust
*Added support for fetching data from /api/info service
*Cleaned up JSON strings in the unit tests to build JSON Objects; seemed cleaner to read

Note: I kept running into white space goofyness so I converted all tabs to spaces.

### CODE REVIEW

File: kount-access-java-sdk/src/main/java/com/kount/kountaccess/AccessSdk.java [hunk 1]
```diff
@@ -43,541 +44,800 @@
...
+
+    private void verifyUniq(String uniq) throws AccessException {
+        if (uniq.length() > 32) {
+            throw new AccessException(AccessErrorType.INVALID_DATA,
+                    "Invalid uniq value (" + uniq + ").  Must not exceed 32 characters in length");
+        }
+    }
+
+    private void verifySessionId(String session) throws AccessException {
+        if (null == session || session.length() != 32) {
```
Comment (boyankukushev) [inline kount-access-java-sdk/src/main/java/com/kount/kountaccess/AccessSdk.java]: I can see several identical validations ( `!= null && .length() == 32` -- a single method could be used (although having several methods looks better semantically)

File: kount-access-java-sdk/src/main/java/com/kount/kountaccess/AccessSdk.java [hunk 2]
```diff
@@ -43,541 +44,800 @@
...
+     * @param value
+     *            The String to convert
+     * @return The converted string.
+     */
+    private String hashValue(String value) {
+        // Don't do anything if the value is empty
+        if (null == value || value.isEmpty()) {
+            return null;
+        }
+        MessageDigest md;
```
Comment (boyankukushev) [inline kount-access-java-sdk/src/main/java/com/kount/kountaccess/AccessSdk.java]: If everybody around is OK with adding one more dependency -- [`Apache Common Codec`](http://commons.apache.org/proper/commons-codec/) and then:
`String sha256hex = org.apache.commons.codec.digest.DigestUtils.sha256Hex(stringText);`

File: kount-access-java-sdk/src/main/java/com/kount/kountaccess/AccessSdk.java [hunk 3]
```diff
@@ -43,541 +44,800 @@
...
+
+        return null;
+    }
+
+    /**
+     * Handles the get request for the device info
+     */
+    private String getRequest(String urlString) throws AccessException {
+        try {
+            CloseableHttpClient client = getHttpClient();
```
Comment (boyankukushev) [inline kount-access-java-sdk/src/main/java/com/kount/kountaccess/AccessSdk.java]: I can't find closing the client ... either create the object in the try-with-resources
`try (CloseableHttpClient client = getHttpClient()) { ... }`
or add a `finally` at the end.

File: kount-access-java-sdk/src/main/java/com/kount/kountaccess/AccessSdk.java [hunk 4]
```diff
@@ -43,541 +44,800 @@
...
+     *            The URL to post to
+     * @param values
+     *            The Form parameters
+     * @return The Response as a String.
+     * @throws AccessException
+     *             Thrown if the URL is bad or we can't connect or parse the response.
+     */
+    private String postRequest(String urlString, List<NameValuePair> values) throws AccessException {
+        try {
+            CloseableHttpClient client = getHttpClient();
```
Comment (boyankukushev) [inline kount-access-java-sdk/src/main/java/com/kount/kountaccess/AccessSdk.java]: Same the one for `getRequest`

File: kount-access-java-sdk/src/main/java/com/kount/kountaccess/AccessSdk.java (hunk 2 above)
Comment (boyankukushev) [reply kount-access-java-sdk/src/main/java/com/kount/kountaccess/AccessSdk.java]: Just checked -- we have commons-codec dependency already.

File: kount-access-java-sdk/src/main/java/com/kount/kountaccess/AccessSdk.java [hunk 5]
```diff
@@ -43,541 +44,800 @@
...
+
+    private void verifyDeviceId(String deviceId) throws AccessException {
+        if (null == deviceId || deviceId.length() != 32) {
+            throw new AccessException(AccessErrorType.INVALID_DATA,
+                    String.format("Invalid deviceId ().  Must be 32 characters in length", deviceId));
+        }
+    }
+
+    private void verifyTrustState(String trustState) throws AccessException {
+        final String validTDI[] = {"trusted", "banned", "not_trusted" };
```
[the remark this answers is not in the document]
Comment (cmavromichalis) [reply kount-access-java-sdk/src/main/java/com/kount/kountaccess/AccessSdk.java]: This is much more readable, good catch!

File: kount-access-java-sdk/src/main/java/com/kount/kountaccess/AccessSdk.java (hunk 1 above)
Comment (cmavromichalis) [reply kount-access-java-sdk/src/main/java/com/kount/kountaccess/AccessSdk.java]: I think I would prefer to maintain the readability and not create a method. I can be convinced though if it's something you feel strongly on changing.

Comment (boyankukushev) [reply kount-access-java-sdk/src/main/java/com/kount/kountaccess/AccessSdk.java]: No, separate methods are OK.

File: kount-access-java-sdk/src/main/java/com/kount/kountaccess/AccessSdk.java (hunk 3 above)
Comment (boyankukushev) [reply kount-access-java-sdk/src/main/java/com/kount/kountaccess/AccessSdk.java]: Sorry for this one; got mislead by the whitespace diff and it looks like it's an inherited issue. Could you still fix this? Thanks.

Comment (cmavromichalis) [reply kount-access-java-sdk/src/main/java/com/kount/kountaccess/AccessSdk.java]: Yeah no problem; will fix.

File: kount-access-java-sdk/src/main/java/com/kount/kountaccess/AccessSdk.java [hunk 6]
```diff
@@ -43,541 +44,800 @@
...
+                    String.format("Invalid returnValue (%s).  Must be an integer between %d and %d", returnValue, bitMin, bitMax));
+        }
+    }
+
+    // Check we have the right parameters passed for the information requested.
+    private void verifyGatherInfoParameters(String session, String returnValue, String deviceId, String uniq) throws AccessException {
+        int deviceInfoBits   = 0b0001;
+        int velocityBits     = 0b0010;
+        int thresholdBits    = 0b0100;
+        int trustedStateBits = 0b1000;
```
[the remark this answers is not in the document]
Comment (cmavromichalis) [reply kount-access-java-sdk/src/main/java/com/kount/kountaccess/AccessSdk.java]: Fixed; I placed them before the function they are used in. I suppose the code style is typically final constants at the top but as this code is quite large I thought before the function is OK. 

What do you think is best?

[the remark this answers is not in the document]
Comment (boyankukushev) [reply kount-access-java-sdk/src/main/java/com/kount/kountaccess/AccessSdk.java]: Looks good.

**Comment by boyankukushev** [comment]
Sorry guys, I didn't understand that this is still a work in progress.

Wouldn't it be better to just create smaller pull requests for each new feature/issue you want to implement?

**Comment by cmavromichalis** [comment]
@boyankukushev I failed to communicate it should not be merged to master yet, sorry about that! We would still like this in master but not now. 

And yes on the smaller pull requests. The maven work should have been separate. When we resubmit this for merge I will break that work out.

**Pull request merged**